src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,141 @@
+from collections import deque
+import re
 from typing import Optional, Type
 
+import ape
 from ape.exceptions import ContractLogicError, TransactionError
 
 
 class RevertsContextManager:
-    def __init__(self, expected_message: Optional[str] = None):
+    def __init__(self, expected_message: Optional[str] = None, dev_message: Optional[str] = None):
+        if dev_message is not None and expected_message is not None:
+            raise ValueError("Cannot supply both 'expected_message' and 'dev_message'.")
+
         self.expected_message = expected_message
+        self.dev_message = dev_message
 
-    def __enter__(self):
-        pass
+    def _check_dev_message(self, exc: ContractLogicError):
+        """
+        Attempts to extract a dev-message from the contract source code by inspecting what
+        instruction(s) led to a transaction revert.
 
-    def __exit__(self, exc_type: Type, exc_value: Exception, traceback) -> bool:
-        if exc_type is None:
-            raise AssertionError("Transaction did not revert.")
+        Raises:
+            AssertionError: When the trace or source can not be retrieved, the trace does not end in
+            a REVERT instruction, or the dev message cannot be found or does not match the expected
+            dev message.
+        """
+        if exc.txn is None:
+            raise AssertionError("Could not fetch transaction information to check dev message.")
 
-        if not isinstance(exc_value, ContractLogicError):
+        receipt = exc.txn.provider.get_receipt(
+            exc.txn.txn_hash,
+            raise_on_fail=False,
+        )
+
+        try:
+            contract = ape.Contract(receipt.receiver)
+        except ValueError as exc:
             raise AssertionError(
-                f"Transaction did not revert.\n"
-                f"However, an exception of type {type(exc_value)} occurred: {exc_value}"
-            ) from exc_value
+                f"Could not fetch contract at {receipt.receiver} to check dev message."
+            ) from exc
+
+        if contract.contract_type.pcmap is None:
+            raise AssertionError("Compiler does not support source code mapping.")
+
+        try:
+            trace = deque(exc.trace)
```

Comment:
@unparalleled-js @NotPeopling2day @fubuloubu 

Depending on the contract and transaction being made, you could end up with a very long trace

I'm not fully utilizing `deque`'s features right now but one of them is that it provides a `maxlen`. Is this something that we should implement and limit trace traversal to a certain number of instructions or can we just allow to examining an entire trace to try and find a dev message?

Unfortunately we do have to fetch the whole trace since we start at the bottom and work backwards. IIRC the EVM trace does like a streaming fetch - we won't get that benefit since we have to grab all of it first.

Reply:
For display purposes, could we only show the last _n_ lines unless the user does `-v debug` in which case, show the full trace? Where `n` is like 10 or 15.